File: FrontEnd/my-app/src/Interfaces/course.ts
import type { EnrolledCourseResponse } from "./enrolledCourse";

export interface course{
    id:number;
    name:string;
    description:string;
    department:string;
    courseCode:string;
    startDate:Date;
    endDate:Date;
    teacherName:string;
    credits:number;
    maxStudents:number;
    enrolledStudents:number;
}

export interface CourseRequest{
    name:string;
    description:string;
    courseCode:string;
    startDate:Date;
    endDate:Date;
    departmentName:string;
    teacherUserName:string;
    creditHours:number;
    maxStudents:number;
}

export interface CourseSidebar {
courseCode:string;
creditHours:number;
startDate:Date;
endDate:Date;
enrolledStudents:number;
name:string;
teacherUserName:string;
}

File: Backend/Uni/src/main/java/UnitSystem/demo/DataAccessLayer/Entities/Course.java
package UnitSystem.demo.DataAccessLayer.Entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

@Entity
@Getter
@Setter
@Builder
@Table(name = "courses")
@RequiredArgsConstructor
@AllArgsConstructor
public class Course {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "course_name", unique = true)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "course_dep", nullable = false)
    private Department department;

    @Column(name = "credits")
    private int credits;

    @Column(name = "course_description")
    private String description;

    @Column(name = "capacity")
    private int capacity;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "course_code", unique = true)
    private String courseCode;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "teacher_id", nullable = false)
    private Teacher teacher;

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EnrolledCourse> courseEnrollments = new HashSet<>();

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<Announcement> announcements = new HashSet<>();

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    @JsonIgnore
    private Set<Message> messages = new HashSet<>();
}
